### PR TITLE
porting in DelimiterBasedFrameDecoder

### PR DESCRIPTION
**Pull request by jeoffman**

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@jeoffman__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    ///   | ABC | DEF |
+    ///  +-----+-----+
+    ///  rather than incorrectly choosing '\r\n' as the first delimiter:
+    ///  +----------+
+    ///  | ABC\nDEF |
+    ///  +----------+
+    /// </summary>
+    public class DelimiterBasedFrameDecoder : ByteToMessageDecoder
+    {
+        private IByteBuffer[] delimiters;
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: code style: do not specify default visibility (private on type members, internal on types)

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+            IByteBuffer a = delimiters[0];
+            IByteBuffer b = delimiters[1];
+            if (a.Capacity < b.Capacity)
+            {
+                a = delimiters[1];
+                b = delimiters[0];
+            }
+            return a.Capacity == 2 && b.Capacity == 1
+                    && a.GetByte(0) == '\r' && a.GetByte(1) == '\n'
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: code style: use single indentation

File: src/DotNetty.Codecs/DotNetty.Codecs.csproj [hunk 3]
```diff
@@ -84,7 +86,5 @@
   <Target Name="AfterBuild">
   </Target>
   -->
-  <ItemGroup>
-    <Folder Include="Strings\" />
-  </ItemGroup>
```
Comment (nayato) [inline src/DotNetty.Codecs/DotNetty.Codecs.csproj]: can you pls verify that it doesn't affect placement of Strings\StringEncoder.cs and Strings\StringDecoder.cs in project structure? Should be under Strings Folder in project.

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            : this(maxFrameLength, true, failFast, new IByteBuffer[] { delimiter })
+        {
+        }
+
+        public DelimiterBasedFrameDecoder(int maxFrameLength, IByteBuffer[] delimiters)
+            : this(maxFrameLength, true, true, delimiters )
+        {
+        }
+
+        public DelimiterBasedFrameDecoder(int maxFrameLength, bool stripDelimiter, IByteBuffer[] delimiters)
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: `params IByteBuffer[] delimeters`. Same for other constructors accepting byte buffer array

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            : this(maxFrameLength, stripDelimiter, true, delimiters)
+        {
+        }
+
+        /// <summary>Returns true if the delimiters are "\n" and "\r\n"</summary>
+        private static bool IsLineBased(IByteBuffer[] delimiters)
+        {
+            if (delimiters.Length != 2)
+                return false;
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: enclose in curly braces

File: src/DotNetty.Codecs/Delimiters.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+﻿using DotNetty.Buffers;
```
Comment (nayato) [inline src/DotNetty.Codecs/Delimiters.cs]: copyright headr

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 7]
```diff
@@ -0,0 +1,286 @@
+﻿using System;
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: copyright header

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        }
+
+        /// <summary>Create a frame out of the <see cref="DotNetty.Buffers.IByteBuffer"/> and return it</summary>
+        /// <param name="ctx">the <see cref="DotNetty.Transport.Channels.IChannelHandlerContext" /> which this <see cref="DotNetty.Codecs.ByteToMessageDecoder"/> belongs to</param>
+        /// <param name="buffer">the <see cref="DotNetty.Buffers.IByteBuffer" /> from which to read data</param>
+        /// <returns>the <see cref="DotNetty.Buffers.IByteBuffer" /> which represent the frame or null if no frame could be created.</returns>
+        protected object Decode(IChannelHandlerContext ctx, IByteBuffer buffer)
+        {
+            if (lineBasedDecoder != null)
+                return lineBasedDecoder.Decode(ctx, buffer);
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: code style: enclose in curly braces

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+            // Try all delimiters and choose the delimiter which yields the shortest frame.
+            int minFrameLength = int.MaxValue;
+            IByteBuffer minDelim = null;
+            foreach(IByteBuffer delim in delimiters)
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: code style: missing space: `foreach (`

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+                        Fail(tooLongFrameLength);
+                    }
+                    return null;
+                }
+
+                if (minFrameLength > maxFrameLength)
+                {
+                    // Discard read frame.
+                    buffer.SkipBytes(minFrameLength + minDelimLength);
+                    Fail(minFrameLength);
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: code style: missing `this.`. Pls check for other such occasions.

File: src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+         * the first needle found in the haystack.  -1 is returned if no needle is
+         * found in the haystack.
+         */
+        private static int IndexOf(IByteBuffer haystack, IByteBuffer needle)
+        {
+            for (int i = haystack.ReaderIndex; i < haystack.WriterIndex; i ++)
+            {
+                int haystackIndex = i;
+                int needleIndex;
+                for (needleIndex = 0; needleIndex < needle.Capacity; needleIndex ++)
```
Comment (nayato) [inline src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs]: code style: unnecessary space, should be `needleIndex++`

**Comment by nayato** [comment]
Please fix formatting (per [Code Style](https://github.com/Azure/DotNetty/wiki/C%23-Coding-Style)) + other small comments. Looks good otherwise. Once done, pls squash commits and try following these recommendations for commit message: http://netty.io/wiki/writing-a-commit-message.html.
And of course, please sign CLA :)

**Comment by azurecla** [comment]
__@jeoffman__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

**Comment by nayato** [comment]
@jeoffman, looks good. Please rebase to latest version and squash. I'll merge once you're done with this. Thanks!

**Comment by jeoffman** [comment]
Tried to squash using SourceTree - my first so I probably botched it. Please let me know.

**Pull request closed without merging**